pubtools/pulplib/_impl/model/maintenance.py

```diff
@@ -0,0 +1,193 @@
+        except jsonschema.exceptions.ValidationError as error:
+            LOG.exception("%s.from_data invoked with invalid data", cls.__name__)
+            raise InvalidDataException(str(error))
+
+        entries = []
+        for repo_id, details in data["repos"].items():
+            entries.append(MaintenanceEntry(repository_id=repo_id, **details))
+
+        maintenance = cls(
+            last_updated=data["last_updated"],
+            last_updated_by=data["last_updated_by"],
+            entries=tuple(entries),
+        )
+
+        return maintenance
+
+    def _json(self):
+        """Output a json format report.
+
+        Returns:
+            A string contains the maintenance report in json format.
+        """
+        report = {
+            "last_updated": self.last_updated,
+            "last_updated_by": self.last_updated_by,
+            "repos": {},
+        }
+
+        for entry in self.entries:
+            report["repos"].update(
+                {
+                    entry.repository_id: {
+                        "message": entry.message,
+                        "owner": entry.owner,
+                        "started": entry.started,
+                    }
+                }
+            )
+
+        return json.dumps(report, indent=4, sort_keys=True)
+
+    def add(self, repo_ids, **kwargs):
+        """Add entries to maintenance report and update the timestamp. Every
+        entry added to the report represents a repository in maintenace mode.
+
+        Args:
+            repo_ids (list[str]):
+                A list of repository ids. New entries with these repository ids will
+                be added to the maintenance report.
+
+                Note: it's users' responsibility to make sure the repository exists in
+                the pulp server, this method doesn't check the repository's existence.
+
+            Optional keyword args:
+                message (str):
+                    Reason why put the repo to maintenance.
+                owner (str):
+                    Who set the maintenance mode.
+        Returns:
+           :class:`~pubtools.pulplib.MaintenanceReport`
+                A copy of this maintenance report with added repositories.
+
+        """
+        message = kwargs.get("message") or "Maintenance mode is enabled"
+        owner = kwargs.get("owner") or self._OWNER
+
+        to_add = []
+        for repo in repo_ids:
+            to_add.append(
+                MaintenanceEntry(repository_id=repo, owner=owner, message=message)
+            )
+        entries = list(self.entries)
+        entries.extend(to_add)
+
+        report = attr.evolve(
+            self,
+            entries=tuple(entries),
+            last_updated_by=owner,
+            last_updated=iso_time_now(),
+        )
+
+        # filter out duplicate entries by export json
+        return MaintenanceReport._from_data(json.loads(report._json()))
```

Comment:
OK, I should have noticed this earlier, this is showing an odd design mismatch now:

Why is it that, when deserializing from JSON to a MaintenanceReport, from_data accepts a raw dict and so the job of decoding JSON into a dict lives elsewhere... but when *serializing*, MaintenanceReport doesn't produce a raw dict but instead produces a JSON string?  It seems like an odd inconsistency.  Would it be better if it both accepted and produced raw dicts, instead of accepting a raw dict but producing a JSON string?

This isn't really important though, given that both are now private API.